plasmapy/diagnostics/thomson.py

```diff
@@ -204,114 +444,422 @@ def spectral_density(
+    if instr_func is not None and callable(instr_func):
+
+        # Create an array of wavelengths of the same size as wavelengths
+        # but centered on zero
+        wspan = (np.max(wavelengths) - np.min(wavelengths)) / 2
+        eval_w = np.linspace(-wspan, wspan, num=wavelengths.size)
+        instr_func_arr = instr_func(eval_w)
+
+        if type(instr_func_arr) != np.ndarray:
+            raise ValueError(
+                "instr_func must be a function that returns a "
+                "np.ndarray, but the provided function returns "
+                f" a {type(instr_func_arr)}"
+            )
+
+        if wavelengths.shape != instr_func_arr.shape:
+            raise ValueError(
+                "The shape of the array returned from the "
+                f"instr_func ({instr_func_arr.shape}) "
+                "does not match the shape of the wavelengths "
+                f"array ({wavelengths.shape})."
+            )
+
+        instr_func_arr *= 1 / np.sum(instr_func_arr)
+    else:
+        instr_func_arr = None
+
+    alpha, Skw = spectral_density_lite(
+        wavelengths.to(u.m).value,
+        probe_wavelength.to(u.m).value,
+        n.to(u.m ** -3).value,
+        Te.to(u.K).value,
+        Ti.to(u.K).value,
+        efract=efract,
+        ifract=ifract,
+        ion_z=ion_z,
+        ion_mass=ion_mass.to(u.kg).value,
+        ion_vel=ion_vel.to(u.m / u.s).value,
+        electron_vel=electron_vel.to(u.m / u.s).value,
+        probe_vec=probe_vec,
+        scatter_vec=scatter_vec,
+        instr_func_arr=instr_func_arr,
+    )
+
+    return alpha, Skw * u.s / u.rad
+
+
+# ***************************************************************************
+# These functions are necessary to interface scalar Parameter objects with
+# the array inputs of spectral_density
+# ***************************************************************************
+
+
+def _count_populations_in_params(params, prefix):
+    """
+    Counts the number of entries matching the pattern prefix_i in a
+    list of keys
+    """
+    keys = list(params.keys())
+    return len(re.findall(prefix, ",".join(keys)))
 
-    # Calculate plasma parameters
-    vTe = thermal_speed(Te, particle="e-")
-    vTi, ion_z = [], []
-    for T, ion in zip(Ti, ion_species):
-        vTi.append(thermal_speed(T, particle=ion).value)
-        ion_z.append(ion.charge_number * u.dimensionless_unscaled)
-    vTi = vTi * vTe.unit
-    zbar = np.sum(ifract * ion_z)
-    ne = efract * n
-    ni = ifract * n / zbar  # ne/zbar = sum(ni)
-    # wpe is calculated for the entire plasma (all electron populations combined)
-    wpe = plasma_frequency(n=n, particle="e-")
 
-    # Convert wavelengths to angular frequencies (electromagnetic waves, so
-    # phase speed is c)
-    ws = (2 * np.pi * u.rad * C / wavelengths).to(u.rad / u.s)
-    wl = (2 * np.pi * u.rad * C / probe_wavelength).to(u.rad / u.s)
+def _params_to_array(params, prefix, vector=False):
```

Comment:
```suggestion
def _params_to_array(
    params: Dict[str, Any], prefix: str, vector: bool = False
) -> numpy.ndarray:
```

Add annotations.  This will need the same import change suggested in the `_cound_populations_in_params`.